### ISSUE TITLE
Rotation Equivariance not perfect yet

### ISSUE THREAD

**Issue by drewm1980** [deworrall92/groupConvolutions#5]
On your project page and in your youtube video, there are still parts of the image that are clearly not equivariant to rotation; visually speaking they swing all the way from black to white.

Have you figured out what causes that yet? 

Is this a side-effect of rotational equivariance only being enforced at the patch centers? 

Or maybe it's some aliasing thing caused by the square window and your windowing function needs to go to hard zero beyond a certain radius? 

Maybe using convolutions larger than 3x3 would clean it up?

I'm still studying up on steerable filters to try to understand how your network works...

**Comment by danielewworrall** [comment]
So I'm not 100% sure as of yet, since I'm currently out of the office for a few weeks. There are two possible sources for this artefact, which come to mind.

1) Most likely is that when we motion correct, we have to unrotate the sampled activations back into a coordinate frame aligned with the original pre-rotated image. This leads to some aliasing, which is the twinkling you see in the video and GIFs. We've been wondering how to correct this, perhaps a pre-filtering step will be needed, but I'm not sure as of yet.

2) Certain architectures may be more unstable under high-frequency content in the images. Bruna and Mallat state this more technically as the Fourier transform being unstable to high-frequency distortions in the input signal. Our convolutions are essentially a windowed Fourier transform in orientation and so if you're not careful the high frequency information may make activations at a certain depth flicker. Now I'm not sure whether this is really happening, and under the assumption that equivariance is preferable to non-equivariance, I'd assume that high frequency instability would be 'learned away' when we fit the weights during training. 

I'd say this second explanation is maybe somewhat tenuous compared to my first, but I cannot rule anything out until I've had a closer look.

**Comment by StephanGarbin** [comment]
I'm closing this as there is no actual issue relating to the code here, which is in alignment with the publication.

**Comment by drewm1980** [comment]
Hi Daniel / @deworrall92 , thanks for the response!

I looked at the animation, and unfortunately the issue isn't just in the final motion correction you do to make the stable animation.  If you look at the un-motion corrected output, you can see that it has the same artifact.

I find it most visible if you look at the youtube video at fullscreen, at 2:29 where you show the (rotating) output of your filter:

https://youtu.be/qoWAFBYOtoU?t=149

The point that sticks out to my eyes the most is a point in the fur above the girl's left eye that cycles from white to black and back every 90 degrees.

If I understand correctly, that probably rules out your theory (1), though there could still be a mistake in how the animation is generated, even before the stabilization filter (hope so).

The fact that the artifact has more than "m=1" periodicity is a clue.

@StephanGarbin I'd keep this open until you guys figure it out. I think it's premature to declare that the problem is with the formulation in the publication; I'm still hopeful it's something trivial. If you want to document the "paper version" for posterity, you could make a tagged version. Then you're free to post improved implementations, fix bugs that might affect experimental results, etc... in the master branch.

**Comment by danielewworrall** [comment]
Hey I hope it's a bug as you say. Maybe we got a sign flipped somewhere and applied a m=-1 convolution instead of a m=1 convolution. Otherwise I'll have to get my hands dirty. Thanks for giving this so much attention.

For the meantime I shan't be able to work on this until the end of next week, when I return to the office (and a computer). So sit tight and I'll let you know in due course.

**Comment by StephanGarbin** [comment]
I'm re-opening this after examining the video as pointed out above - There are definitely some artifacts.

@drewm1980 : Thanks for linking the video, it's always easier to see with a concrete pointer after you've been looking at it for as long as we have.
We are aware of github's functionality and will release accordingly should that be necessary. Thanks for your input, it's great to see people interested in this work :)

**Comment by drewm1980** [comment]
A hunch I had was that maybe your artifact was coming from the samples in the corner of your filter, and you could try clamping them to zero. I deleted it for brevity, but it just popped up in this preprint on "Oriented Response Networks":

https://arxiv.org/pdf/1701.01833v1.pdf

"Note that points outside the inscribed circle are padded with 0."

They are also working with small convolutional windows and heavily quantized rotations... they don't post an animation like yours, so it's hard to say how bad their artifacts are, but it would be interesting if someone did a comparison to sanity check how well all of these concurrent papers claiming rotation equivariance are actually achieving it. Seems you have tooling for tests like this in your repo, so maybe your group is the right one to do that. Update: They say they'll post their code after the cvpr review process is done.

For "good to within 45 degrees" is probably good enough for a lot of applications, i.e. classification, but that isn't going to cut it for a lot of orientation measuring applications, i.e. robotic grasping in pick and place.

**Comment by nthomas103** [comment]
Any update on this issue?

Here is a simple test of hn_lite.conv2d, using randomly initialized weight matrices with your harmonic network functions (that is, with no training):

https://gist.github.com/nthomas103/4edaddbb81ee4e90d162c81bf7b7d148
(written for use in an IPython notebook)

It works fine for stride 1 and less than 4 layers, but, for example, 8 layers & stride 1 or 2 layers and stride 2 both show significant discrepancies.

Have you seen something similar in your tests?

**Comment by danielewworrall** [comment]
Hey guys. 

So just to let you know, there is a [new version of the paper online](https://arxiv.org/pdf/1612.04642.pdf). In it we have added some minor details about the new sampling strategy and anti-aliasing, that we've added. 

As for nthomas103's experiments. Thanks for the script. I ran it and it's a nice validation of the stability of the hn_lite convolution for networks, which aren't too deep. Beyond 5 layers, I agree with your conclusions that things start to break down. I have a few comments on this, some in my favour and some which aren't and need further investigation/resolving.

1. **Edge effects**: The receptive field of a 3x3 convolution at 5 layers is 11x11, so measuring the invariance properties on MNIST (28x28) is not the best idea because of windowing effects. I got better results with larger images. Maybe you could zero pad the input a lot, if you want to work on small images.
2. **Initialisation**: Seeing that the script runs on randomly initialised weight matrices, the activations either blow up/vanish without, say, training time batch normalisation or judicious tinkering with the standard deviation of the initialization. This shouldn't theoretically change the invariance properties of the convolution, but can lead to difficult-to-interpret blobby outputs.
3. **Strides**: I haven't worked out what's going on here yet--I will soon enough. My guesses/inferences are 
**A.** be careful with stride sizes, we've found that if you are not, the centres of equivariance will shift wrt the image under rotation. For instance, with a 28x28 input and a 3x3 filter (we now support even kernel sizes btw :)), and stride 2, the centre of the top-left filter will be in pixel (0,0), and the centre of the bottom-right will be (26,26), not (27,27), which is the true bottom-right pixel. This means that a rotation of the image, leads to a rotation + translation. You could fix this with appropriate padding/filter sizes for the centres of equivariance to align with the centre of input rotation.
**B.** Strides lead to subsampling of the input signal, and without proper anti-aliasing (just blur the image a bit with a Gaussian), you'll get spurious aliasing artifacts. This is why we used mean-pooling in the paper to downsample. Conventional CNNs do not use this, and it seems they just learn the anti-aliasing by brute force. This is what we found as well, that training led to anti-aliasing...although it would be nice to have it included from random weights as well.
4. I've started running experiments on the stability of the feature maps under rotation for CNNs of depth >= 4. I measure mean discrepancy as the root mean squared error of a rotation-corrected feature map, versus the unrotated version. Indeed there is a cyclic discrepancy. I don't know the source as of yet, but am looking into it right now.
![discrepancy](https://cloud.githubusercontent.com/assets/7756208/25046279/7cd771ee-2128-11e7-8e8f-84bf5648b8ef.png)

**Comment by nthomas103** [comment]
Thanks for the detailed response!
